### DISCUSSION TITLE
sBTC UTXO Consolidation

### DISCUSSION THREAD

**Discussion by 8marz8** [stacks-sbtc/sbtc#10] [Research]
Some UTXO consolidation details were laid out during our HLD discussion. Recording them here for better reference and future LLD/Implementation details for UTXO consolidation that needs to be carried out by the signers.

1. Collect all deposits
2. Collect all withdraws
3. Start building BTC tx
4. Add signer UTXO as input
5. Add all deposit UTXOs as inputs
6. Add all withdraws as outputs
7. Add change output as new signer UTXO
8. Broadcast BTC Tx
10. Call mint in sBTC with:
    * list of (addr, recipient)UTXO.
    * BTC TXID (for visibility)
11. Call sBTC function to handle withdraws

**Comment by netrome** [comment]
For reference, the details of the BTC transaction are under [this discussion](https://github.com/Trust-Machines/sbtc-v1/discussions/6).

**Comment by AshtonStephens** [comment]
Ensure that the signer are the only ones able to spend in the block that has the consolidation occur.

**Reply by netrome** [reply]
It would be nice if we could make this an optional behavior in the Signer. 

Optimistic mode -> no validation needed.
Pessimistic mode -> reject any deposit request which do not explicitly prove that the signers are the only entity which can spend this.

We'd default to running optimistically, but switch to pessimistic if any deposit UTXOs turn out to be already spent.
